server/src/main/java/org/apache/iotdb/db/engine/storagegroup/virtualSg/VirtualStorageGroupManager.java

```diff
@@ -0,0 +1,136 @@
+  }
+
+  /**
+   * push forceCloseAllWorkingTsFileProcessors down to all sg
+   */
+  public void forceCloseAllWorkingTsFileProcessors() throws TsFileProcessorException {
+    for(StorageGroupProcessor storageGroupProcessor : virtualStorageGroupProcessor){
+      if(storageGroupProcessor != null){
+        storageGroupProcessor.forceCloseAllWorkingTsFileProcessors();
+      }
+    }
+  }
+
+  /**
+   * push syncCloseAllWorkingTsFileProcessors down to all sg
+   */
+  public void syncCloseAllWorkingTsFileProcessors(){
+    for(StorageGroupProcessor storageGroupProcessor : virtualStorageGroupProcessor){
+      if(storageGroupProcessor != null){
+        storageGroupProcessor.syncCloseAllWorkingTsFileProcessors();
+      }
+    }
+  }
+
+  /**
+   * push check ttl down to all sg
+   */
+  public void checkTTL(){
+    for(StorageGroupProcessor storageGroupProcessor : virtualStorageGroupProcessor){
+      if(storageGroupProcessor != null){
+        storageGroupProcessor.checkFilesTTL();
+      }
+    }
+  }
+
+  /**
+   * get processor from device id
+   * @param partialPath device path
+   * @return virtual storage group processor
+   */
+  @SuppressWarnings("java:S2445")
+  // actually storageGroupMNode is a unique object on the mtree, synchronize it is reasonable
+  public StorageGroupProcessor getProcessor(PartialPath partialPath, StorageGroupMNode storageGroupMNode)
+      throws StorageGroupProcessorException, StorageEngineException {
+    int loc = partitioner.deviceToStorageGroup(partialPath);
+
+    StorageGroupProcessor processor = virtualStorageGroupProcessor[loc];
+    if (processor == null) {
+      // if finish recover
+      if (StorageEngine.getInstance().isAllSgReady()) {
+        synchronized (storageGroupMNode) {
+          processor = virtualStorageGroupProcessor[loc];
+          if (processor == null) {
+            processor = StorageEngine.getInstance()
+                .buildNewStorageGroupProcessor(storageGroupMNode.getPartialPath(), storageGroupMNode, String.valueOf(loc));
+            virtualStorageGroupProcessor[loc] = processor;
+          }
+        }
+      } else {
+        // not finished recover, refuse the request
+        throw new StorageEngineException(
+            "the sg " + partialPath + " may not ready now, please wait and retry later",
+            TSStatusCode.STORAGE_GROUP_NOT_READY.getStatusCode());
+      }
+    }
+
+    return processor;
+  }
+
+  /**
+   * recover
+   * @param storageGroupMNode logical sg mnode
+   */
+  public void recover(StorageGroupMNode storageGroupMNode) throws StorageGroupProcessorException {
+    for (int i = 0; i < partitioner.getPartitionCount(); i++) {
+      StorageGroupProcessor processor = StorageEngine.getInstance()
+          .buildNewStorageGroupProcessor(storageGroupMNode.getPartialPath(), storageGroupMNode, String.valueOf(i));
+      virtualStorageGroupProcessor[i] = processor;
```

Comment:
Maybe different partitions can be recovered in parallel.

Reply:
fixed~